Signal IsCompleted from Response.BodyWriter after abort

senbar opened the pull request:
# After RequestAborted, Response.BodyWriter does not signal IsCompleted/IsCanceled

<!-- Thank you for submitting a pull request to our repo. -->

<!-- If this is your first PR in the ASP.NET Core repo, please run through the checklist
below to ensure a smooth review and merge process for your PR. -->
<!-- Once all that is done, you're ready to go. Open the PR with the content below. -->
**In Progress!** - IIS fix in progress
Change logic returning FlushResult when interacting with OutputProducer to be more consistent. 
For now Changes fix only Kestrel case, with IIS I have problems debugging and can't find routine responsible for creating FlushResults


## Description

Add logic for keeping track of if output producer pending flush was canceled. Modify FlushResult returned by FlushAsync to retun IsCompleted and IsCanceled for if stream was completed and stream had pending flush canceled respectively. I implemented it on http 1 2 3 and have written tests for them. I am not sure if tests are in the correct classes so feel free to correct me here.

Fixes #32331

senbar committed:
After RequestAborted, Response.BodyWriter does not signal IsCompleted
* added test for http1 http2 and http3 cases for kestrel
* changed flush logic to return FlushResult with IsCompleted set to true on aborted stream

senbar committed:
After RequestAborted, Response.BodyWriter does not signal IsCompleted/IsCanceled (#32331)
    * fix test logic to check for IsCompleted for aborted case
    * add tests checking CancelPendingFlush logic
    * add private field isPendingFlushCanceled allowing flushing to check if pending flush was canceled
    * fix previous tests naming

senbar committed:
After RequestAborted, Response.BodyWriter does not signal IsCompleted/IsCanceled (#32331)
    * fix test logic to check for IsCompleted for aborted case
    * add tests checking CancelPendingFlush logic
    * add private field isPendingFlushCanceled allowing flushing to check if pending flush was canceled
    * fix previous tests naming

senbar commented:
> This looks good at first glance. Why is this still a draft?

I think this only fixes the issue in kestrel case. Or does ISS express use kestrel namespace because to be honest I couldn't find the code responsible for writing to response stream in IIS namespace

halter73 commented:
Here's probably the code you should be looking at for IIS:

- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs#L155-L215
- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/IIS/IIS/src/Core/IISHttpContext.IO.cs#L279-L286
- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/IIS/IIS/src/Core/OutputProducer.cs#L53-L67

I haven't looked into what changes actually need to be made. If you get stuck on the IIS stuff, I'd be happy to just take the Kestrel stuff for now.

Thanks!

halter73 commented:
I think I see the issue. The problem is that IIS is still using a `StreamPipeWriter` which basically doesn't support `IsCompleted`. We might be able to fix this with a change to `StreamPipeWriter` in the https://github.com/dotnet/runtime repo. Or maybe we could consider exposing the `OutputProducer`s inner `PipeWriter` more directly. That'd probably be a lot of work though.

https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/IIS/IIS/src/Core/IISHttpContext.FeatureCollection.cs#L185-L196

Is there anything else you think needs done on the Kestrel side before merging this?

halter73 commented on src/Servers/Kestrel/Core/src/Internal/Http/Http1OutputProducer.cs:
```diff
@@ -143,7 +144,11 @@ public ValueTask<FlushResult> FlushAsync(CancellationToken cancellationToken = d
+            }
+            if (_pendingFlushCanceled)
+            {
+                return new ValueTask<FlushResult>(new FlushResult(true, false));
```
Okay, so there is an issue with this. `CancelPendingFlush()` (and `CancelPendingRead()` FWIW) is not terminal. After returning the first `FlushResult`, it should be possible to flush again normally without it being immediately canceled.

Also, `CancelPendingFlush()` can be called before or during an ongoing flush. It should not prevent a flush from happening even if it's called first. It should just cause the flush to immediately return with a cancelled `FlushResult`. Then next call to `FlushAsync()` shouldn't return a cancelled `FlushResult` unless `CancelPendingFlush()` was called again.

I think this part should have been taken care of already by the call to `_pipeWriter.CancelPendingFlush();` Is it possible this was already behaving correctly, but the intended behavior wasn't well understood?

senbar replied:
So i think the problem is that PipeFlusher either forgets that it has been cancelled or doesn't return correctly. I will look into that, converting this PR to draft for now

senbar commented:
No, Kestrel part is implemented correctly I hope.

halter73 commented on src/Servers/Kestrel/Core/test/Http1/Http1OutputProducerTests.cs:
```diff
@@ -54,6 +54,61 @@ public async Task WritesNoopAfterConnectionCloses()
+        // Close
+        socketOutput.CancelPendingFlush();
+
+        await socketOutput.WriteDataAsync(new byte[] { 1, 2, 3, 4 }, default);
+        var cancelResult = await socketOutput.FlushAsync();
```
```suggestion
        var cancelResult = await socketOutput.WriteDataToPipeAsync(new byte[] { 1, 2, 3, 4 }, default);
```

This was effectively double flushing before because it was calling the internal `WriteDataAsync` API that's called by `HttpResponse.Body` rather than `WriteDataToPipeAsync` that's called by `HttpRessponse.BodyWriter`.

I wonder if this is why you weren't seeing canceled `FlushResult`s  before this change. The second flush shouldn't be canceled as I mentioned earlier.

halter73 commented:
I think we should add functional tests to the following classes to verify the end to end.

- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/Kestrel/test/InMemory.FunctionalTests/ResponseTests.cs
- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2StreamTests.cs
- https://github.com/dotnet/aspnetcore/blob/82e6b60edbeb9c247d6e2b82ea69306eefb613f4/src/Servers/Kestrel/test/InMemory.FunctionalTests/Http3/Http3StreamTests.cs

senbar committed:
Update src/Servers/Kestrel/Core/test/Http1/Http1OutputProducerTests.cs

Co-authored-by: Stephen Halter <halter73@gmail.com>

senbar committed:
Merge branch 'BodyWriter_signal_IsCanceled' of https://github.com/senbar/aspnetcore into BodyWriter_signal_IsCanceled

senbar committed:
Move tests to functional tests classes;
remove all isFlushCancelled logic since cancelling flushing actually was working from the beginning correctly;
change all tests to be using Responose bodywriter interface

halter73 commented on src/Servers/Kestrel/Core/test/Http1/Http1ConnectionTests.cs:
```diff
@@ -696,6 +696,38 @@ public async Task RequestAbortedTokenIsResetBeforeLastWriteAsyncWithContentLengt
+        _http1Connection.Abort(new ConnectionAbortedException());
+        var failResult= await _http1Connection.FlushPipeAsync(new CancellationToken());
+        Assert.True(failResult.IsCompleted);
+    }
```
Nit:

```suggestion
    }

```

Can you also fix the formatting of the `=` signs in `var failResult= await _http1Connection.FlushPipeAsync(new CancellationToken());` and elsewhere?

halter73 commented on src/Servers/Kestrel/Core/test/Http1/Http1OutputProducerTests.cs:
```diff
@@ -54,6 +54,64 @@ public async Task WritesNoopAfterConnectionCloses()
+
+        socketOutput.Pipe.Writer.Complete();
+        socketOutput.Pipe.Reader.Complete();
+    }
```
```suggestion
    }

```

halter73 commented on src/Servers/Kestrel/Core/test/Http1/Http1OutputProducerTests.cs:
```diff
@@ -54,6 +54,64 @@ public async Task WritesNoopAfterConnectionCloses()
+        socketOutput.Dispose();
+
+        await socketOutput.WriteDataAsync(new byte[] { 1, 2, 3, 4 }, default);
+        var cancelResult = await socketOutput.FlushAsync();
+        Assert.True(cancelResult.IsCompleted);
```
```suggestion
        var completedResult = await socketOutput.FlushAsync();
        Assert.True(completedResult.IsCompleted);
```

halter73 reviewed (approved):
The tests look great! Thanks for doing this!

I just have a few formatting nits. There are a few other places in the tests that don't have a space before `=` that I didn't comment on.

halter73 commented:
> Should I leave the connection tests in or delete them?

Sorry for the slow response. Please, leave them. Thanks again @senbar

senbar committed:
rename wrongly named var;
add missing break lines

senbar commented:
Thanks for code review @halter73 :) Fixed cosmetic changes, I am too used to IDE correcting these for me

halter73 commented:
Thanks again @senbar!
